File: maze-walker/python/tests/maze_builder.py
from __future__ import annotations

from maze_walker import (
    Cell,
    CellType,
    Maze,
    NoStartCellException,
    NoEndCellException,
    MultipleStartCellsException,
    MultipleEndCellsException,
)


class MazeBuilder:
    def __init__(self) -> None:
        self._layout = "SE"

    def with_layout(self, layout: str) -> MazeBuilder:
        self._layout = layout
        return self

    def build(self) -> Maze:
        lines = self._layout.split("\n")
        rows = len(lines)
        cols = max(len(line) for line in lines)
        grid: list[list[CellType]] = []

        start: Cell | None = None
        end: Cell | None = None
        start_count = 0
        end_count = 0

        for r in range(rows):
            row: list[CellType] = []
            for c in range(cols):
                ch = lines[r][c] if c < len(lines[r]) else "."
                if ch == "#":
                    row.append(CellType.WALL)
                elif ch == "S":
                    row.append(CellType.START)
                    start = Cell(r, c)
                    start_count += 1
                elif ch == "E":
                    row.append(CellType.END)
                    end = Cell(r, c)
                    end_count += 1
                else:
                    row.append(CellType.OPEN)
            grid.append(row)

        if start_count == 0:
            raise NoStartCellException()
        if start_count > 1:
            raise MultipleStartCellsException()
        if end_count == 0:
            raise NoEndCellException()
        if end_count > 1:
            raise MultipleEndCellsException()

        assert start is not None
        assert end is not None
        return Maze(grid, start, end)

File: maze-walker/typescript/tests/MazeBuilder.ts
import { type Cell } from '../src/Cell.js';
import { CellType } from '../src/CellType.js';
import { Maze } from '../src/Maze.js';
import {
  NoStartCellException,
  NoEndCellException,
  MultipleStartCellsException,
  MultipleEndCellsException,
} from '../src/MazeExceptions.js';

export class MazeBuilder {
  private layout = 'SE';

  withLayout(layout: string): this {
    this.layout = layout;
    return this;
  }

  build(): Maze {
    const lines = this.layout.split('\n');
    const rows = lines.length;
    const cols = Math.max(...lines.map((l) => l.length));
    const grid: CellType[][] = [];

    let start: Cell | undefined;
    let end: Cell | undefined;
    let startCount = 0;
    let endCount = 0;

    for (let r = 0; r < rows; r++) {
      const row: CellType[] = [];
      const line = lines[r]!;
      for (let c = 0; c < cols; c++) {
        const ch = c < line.length ? line[c] : '.';
        switch (ch) {
          case '#':
            row.push(CellType.Wall);
            break;
          case 'S':
            row.push(CellType.Start);
            start = { row: r, col: c };
            startCount++;
            break;
          case 'E':
            row.push(CellType.End);
            end = { row: r, col: c };
            endCount++;
            break;
          default:
            row.push(CellType.Open);
            break;
        }
      }
      grid.push(row);
    }

    if (startCount === 0) throw new NoStartCellException();
    if (startCount > 1) throw new MultipleStartCellsException();
    if (endCount === 0) throw new NoEndCellException();
    if (endCount > 1) throw new MultipleEndCellsException();

    return new Maze(grid, start!, end!);
  }
}
